internal/statestore/backfill.go

```diff
@@ -0,0 +1,152 @@
+package statestore
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gomodule/redigo/redis"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"open-match.dev/open-match/pkg/pb"
+)
+
+// CreateBackfill creates a new Backfill in the state storage. If the id already exists, it will be overwritten.
+func (rb *redisBackend) CreateBackfill(ctx context.Context, backfill *pb.Backfill) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return status.Errorf(codes.Unavailable, "CreateBackfill, id: %s, failed to connect to redis: %v", backfill.GetId(), err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	value, err := proto.Marshal(backfill)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to marshal the backfill proto, id: %s", backfill.GetId())
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	_, err = redisConn.Do("SET", backfill.GetId(), value)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to set the value for backfill, id: %s", backfill.GetId())
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	return nil
+}
+
+// GetBackfill gets the Backfill with the specified id from state storage. This method fails if the Backfill does not exist.
+func (rb *redisBackend) GetBackfill(ctx context.Context, id string) (*pb.Backfill, error) {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "GetBackfill, id: %s, failed to connect to redis: %v", id, err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	return getBackfill(redisConn, id)
+}
+
+// DeleteBackfill removes the Backfill with the specified id from state storage. This method succeeds if the Backfill does not exist.
+func (rb *redisBackend) DeleteBackfill(ctx context.Context, id string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return status.Errorf(codes.Unavailable, "DeleteBackfill, id: %s, failed to connect to redis: %v", id, err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	_, err = redisConn.Do("DEL", id)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to delete the backfill from state storage, id: %s", id)
+		return status.Errorf(codes.Internal, "%v", err)
+	}
+
+	return nil
+}
+
+// UpdateBackfill updates an existing Backfill with new data.
+func (rb *redisBackend) UpdateBackfill(ctx context.Context, backfill *pb.Backfill, updateFunc func(current *pb.Backfill, new *pb.Backfill) (*pb.Backfill, error)) (*pb.Backfill, error) {
+	if updateFunc == nil {
+		return nil, status.Errorf(codes.Internal, "nil updateFunc provided")
+	}
+
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "UpdateBackfill, id: %s, failed to connect to redis: %v", backfill.GetId(), err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	for i := 0; i < 3; i++ {
```

Comment:
I assume we should avoid magic numbers here could go some constant or configurable value.

Reply:
In my opinion, this single case is not worth adding an additional parameter and propagating it everywhere for instance to makefile, helm, or so on. It's really hard to catch and test that scenario live where you can see a big difference in this parameter. But it's better to limit this loop anyway rather than making an infinite one.